src/Halogen/Aff/Driver.purs

```diff
@@ -88,28 +94,52 @@ runUI
   :: forall h r f o eff
    . RenderSpec h r eff
   -> Component h f o (Aff (HalogenEffects eff))
-  -> Aff (HalogenEffects eff) (Driver f eff)
-runUI renderSpec component = _.driver <$> do
+  -> Aff (HalogenEffects eff) (HalogenIO f o (Aff (HalogenEffects eff)))
+runUI renderSpec component = do
   fresh <- liftEff $ newRef 0
-  handleLifecycle \lchs ->
-    runComponent (const (pure unit)) fresh lchs Root component
+  handleLifecycle \lchs -> do
+    listeners <- liftEff $ newRef M.empty
+    runComponent (rootHandler listeners) fresh lchs Root component
       >>= peekVar
       >>= unDriverStateX \st ->
-        -- The record here is a hack around a skolem escape issue. If the typing
-        -- rules for records change so this no longer works it may also be
-        -- fixable with copious type annotations.
-        pure { driver: evalF st.selfRef }
+        pure
+          { query: evalF st.selfRef
+          , subscribe: subscribe fresh listeners
+          }
 
   where
 
+  rootHandler :: Ref (M.Map Int (AV.AVar (Either o Unit))) -> o -> Aff (HalogenEffects eff) Unit
+  rootHandler ref message = do
+    listeners <- liftEff $ readRef ref
+    void $ forkAll $ map (\var -> AV.putVar var (Left message)) listeners
+
+  subscribe
+    :: Ref Int
+    -> Ref (M.Map Int (AV.AVar (Either o Unit)))
+    -> ({ producer :: CR.Producer o (Aff (HalogenEffects eff)) Unit
+        , unsubscribe :: Aff (HalogenEffects eff) Boolean
+        } -> Aff (HalogenEffects eff) Unit)
+    -> Aff (HalogenEffects eff) Unit
+  subscribe fresh ref k = do
+    listenerId <- liftEff $ readRef fresh
+    liftEff $ modifyRef fresh (_ + 1)
```

Comment:
How about
``` purescript
listenerId <- liftEff do
  lid <- readRef fresh
  modifyRef fresh (_ + 1)
  pure lid
```
We get "magic do" optimisations that way? Or I could just leave it 😛